From e5c3d93a6dbb235c9a121bd5c710fae7ddc5811a Mon Sep 17 00:00:00 2001
From: Vincent Bernat <vincent@bernat.im>
Date: Fri, 6 Apr 2018 15:52:48 +0200
Subject: [PATCH 2/3] bgpd: add basic support for ETI and ESI for BGP EVPN

Ethernet Tag ID (ETI) is part of the prefix. It cannot just be ignored
as it needs to be used when checking for prefix uniqueness. Moreover,
when using Quagga as a route reflector, we need to keep its
value. Therefore, we correctly parse and encode it. We also parse
ESI. While not part of the prefix, it needs to be reflected correctly
by Quagga.

Signed-off-by: Vincent Bernat <vincent@bernat.im>
---
 bgpd/bgp_attr_evpn.h |  1 +
 bgpd/bgp_evpn.c      | 61 +++++++++++++++++++++++++++-----------------
 bgpd/bgp_evpn_vty.c  |  4 +--
 3 files changed, 40 insertions(+), 26 deletions(-)

diff --git a/bgpd/bgp_attr_evpn.h b/bgpd/bgp_attr_evpn.h
index 8b55cb30020b..7f7235d0fa77 100644
--- a/bgpd/bgp_attr_evpn.h
+++ b/bgpd/bgp_attr_evpn.h
@@ -37,6 +37,7 @@
 u_long eth_tag_id;
 struct attr;
 
+/* EVPN ESI */
 struct eth_segment_id {
 	u_char val[ESI_LEN];
 };
diff --git a/bgpd/bgp_evpn.c b/bgpd/bgp_evpn.c
index 1ac0d55e6de2..72f3cb829008 100644
--- a/bgpd/bgp_evpn.c
+++ b/bgpd/bgp_evpn.c
@@ -2675,9 +2675,11 @@ static int process_type2_route(struct peer *peer, afi_t afi, safi_t safi,
 {
 	struct prefix_rd prd;
 	struct prefix_evpn p;
+	struct bgp_route_evpn evpn;
 	u_char ipaddr_len;
 	u_char macaddr_len;
 	mpls_label_t *label_pnt;
+	uint32_t eth_tag;
 	int ret;
 
 	/* Type-2 route should be either 33, 37 or 49 bytes or an
@@ -2693,6 +2695,8 @@ static int process_type2_route(struct peer *peer, afi_t afi, safi_t safi,
 		return -1;
 	}
 
+	memset(&evpn, 0, sizeof(evpn));
+
 	/* Make prefix_rd */
 	prd.family = AF_UNSPEC;
 	prd.prefixlen = 64;
@@ -2705,10 +2709,13 @@ static int process_type2_route(struct peer *peer, afi_t afi, safi_t safi,
 	p.prefixlen = EVPN_TYPE_2_ROUTE_PREFIXLEN;
 	p.prefix.route_type = BGP_EVPN_MAC_IP_ROUTE;
 
-	/* Skip over Ethernet Seg Identifier for now. */
-	pfx += 10;
+	/* Copy Ethernet Seg Identifier */
+	memcpy(&evpn.eth_s_id.val, pfx, ESI_LEN);
+	pfx += ESI_LEN;
 
-	/* Skip over Ethernet Tag for now. */
+	/* Copy Ethernet Tag */
+	memcpy(&eth_tag, pfx, 4);
+	p.prefix.eth_tag = ntohl(eth_tag);
 	pfx += 4;
 
 	/* Get the MAC Addr len */
@@ -2753,11 +2760,11 @@ static int process_type2_route(struct peer *peer, afi_t afi, safi_t safi,
 	if (attr)
 		ret = bgp_update(peer, (struct prefix *)&p, addpath_id, attr,
 				 afi, safi, ZEBRA_ROUTE_BGP, BGP_ROUTE_NORMAL,
-				 &prd, label_pnt, 0, NULL);
+				 &prd, label_pnt, 0, &evpn);
 	else
 		ret = bgp_withdraw(peer, (struct prefix *)&p, addpath_id, attr,
 				   afi, safi, ZEBRA_ROUTE_BGP, BGP_ROUTE_NORMAL,
-				   &prd, label_pnt, NULL);
+				   &prd, label_pnt, &evpn);
 	return ret;
 }
 
@@ -2771,6 +2778,7 @@ static int process_type3_route(struct peer *peer, afi_t afi, safi_t safi,
 	struct prefix_rd prd;
 	struct prefix_evpn p;
 	u_char ipaddr_len;
+	uint32_t eth_tag;
 	int ret;
 
 	/* Type-3 route should be either 17 or 29 bytes: RD (8), Eth Tag (4),
@@ -2794,7 +2802,9 @@ static int process_type3_route(struct peer *peer, afi_t afi, safi_t safi,
 	p.prefixlen = EVPN_TYPE_3_ROUTE_PREFIXLEN;
 	p.prefix.route_type = BGP_EVPN_IMET_ROUTE;
 
-	/* Skip over Ethernet Tag for now. */
+	/* Copy Ethernet Tag */
+	memcpy(&eth_tag, pfx, 4);
+	p.prefix.eth_tag = ntohl(eth_tag);
 	pfx += 4;
 
 	/* Get the IP. */
@@ -3425,7 +3435,7 @@ void bgp_evpn_route2json(struct prefix_evpn *p, json_object *json)
 
 	if (p->prefix.route_type == BGP_EVPN_IMET_ROUTE) {
 		json_object_int_add(json, "routeType", p->prefix.route_type);
-		json_object_int_add(json, "ethTag", 0);
+		json_object_int_add(json, "ethTag", p->prefix.eth_tag);
 		json_object_int_add(json, "ipLen",
 				    IS_EVPN_PREFIX_IPADDR_V4(p)
 					    ? IPV4_MAX_BITLEN
@@ -3436,10 +3446,7 @@ void bgp_evpn_route2json(struct prefix_evpn *p, json_object *json)
 		if (IS_EVPN_PREFIX_IPADDR_NONE(p)) {
 			json_object_int_add(json, "routeType",
 					    p->prefix.route_type);
-			json_object_int_add(
-				json, "esi",
-				0); /* TODO: we don't support esi yet */
-			json_object_int_add(json, "ethTag", 0);
+			json_object_int_add(json, "ethTag", p->prefix.eth_tag);
 			json_object_int_add(json, "macLen", 8 * ETH_ALEN);
 			json_object_string_add(json, "mac",
 					       prefix_mac2str(&p->prefix.mac,
@@ -3453,10 +3460,7 @@ void bgp_evpn_route2json(struct prefix_evpn *p, json_object *json)
 
 			json_object_int_add(json, "routeType",
 					    p->prefix.route_type);
-			json_object_int_add(
-				json, "esi",
-				0); /* TODO: we don't support esi yet */
-			json_object_int_add(json, "ethTag", 0);
+			json_object_int_add(json, "ethTag", p->prefix.eth_tag);
 			json_object_int_add(json, "macLen", 8 * ETH_ALEN);
 			json_object_string_add(json, "mac",
 					       prefix_mac2str(&p->prefix.mac,
@@ -3486,14 +3490,17 @@ char *bgp_evpn_route2str(struct prefix_evpn *p, char *buf, int len)
 	char buf2[PREFIX2STR_BUFFER];
 
 	if (p->prefix.route_type == BGP_EVPN_IMET_ROUTE) {
-		snprintf(buf, len, "[%d]:[0]:[%d]:[%s]", p->prefix.route_type,
+		snprintf(buf, len, "[%d]:[%d]:[%d]:[%s]", p->prefix.route_type,
+			 p->prefix.eth_tag,
 			 IS_EVPN_PREFIX_IPADDR_V4(p) ? IPV4_MAX_BITLEN
 						     : IPV6_MAX_BITLEN,
 			 inet_ntoa(p->prefix.ip.ipaddr_v4));
 	} else if (p->prefix.route_type == BGP_EVPN_MAC_IP_ROUTE) {
 		if (IS_EVPN_PREFIX_IPADDR_NONE(p))
-			snprintf(buf, len, "[%d]:[0]:[0]:[%d]:[%s]",
-				 p->prefix.route_type, 8 * ETH_ALEN,
+			snprintf(buf, len, "[%d]:[%d]:[%d]:[%s]",
+				 p->prefix.route_type,
+				 p->prefix.eth_tag,
+				 8 * ETH_ALEN,
 				 prefix_mac2str(&p->prefix.mac, buf1,
 						sizeof(buf1)));
 		else {
@@ -3501,8 +3508,10 @@ char *bgp_evpn_route2str(struct prefix_evpn *p, char *buf, int len)
 
 			family = IS_EVPN_PREFIX_IPADDR_V4(p) ? AF_INET
 							     : AF_INET6;
-			snprintf(buf, len, "[%d]:[0]:[0]:[%d]:[%s]:[%d]:[%s]",
-				 p->prefix.route_type, 8 * ETH_ALEN,
+			snprintf(buf, len, "[%d]:[%d]:[%d]:[%s]:[%d]:[%s]",
+				 p->prefix.route_type,
+				 p->prefix.eth_tag,
+				 8 * ETH_ALEN,
 				 prefix_mac2str(&p->prefix.mac, buf1,
 						sizeof(buf1)),
 				 family == AF_INET ? IPV4_MAX_BITLEN
@@ -3511,8 +3520,9 @@ char *bgp_evpn_route2str(struct prefix_evpn *p, char *buf, int len)
 					   PREFIX2STR_BUFFER));
 		}
 	} else if (p->prefix.route_type == BGP_EVPN_IP_PREFIX_ROUTE) {
-		snprintf(buf, len, "[%d]:[0]:[%d]:[%s]",
+		snprintf(buf, len, "[%d]:[%d]:[%d]:[%s]",
 			 p->prefix.route_type,
+			 p->prefix.eth_tag,
 			 p->prefix.ip_prefix_length,
 			 IS_EVPN_PREFIX_IPADDR_V4(p) ?
 				inet_ntoa(p->prefix.ip.ipaddr_v4) :
@@ -3551,8 +3561,11 @@ void bgp_evpn_encode_prefix(struct stream *s, struct prefix *p,
 			ipa_len = IPV6_MAX_BYTELEN;
 		stream_putc(s, 33 + ipa_len);       // 1 VNI
 		stream_put(s, prd->val, 8);	 /* RD */
-		stream_put(s, 0, 10);		    /* ESI */
-		stream_putl(s, 0);		    /* Ethernet Tag ID */
+		if (attr)
+			stream_put(s, &attr->evpn_overlay.eth_s_id, ESI_LEN);
+		else
+			stream_put(s, 0, 10);
+		stream_putl(s, evp->prefix.eth_tag);	    /* Ethernet Tag ID */
 		stream_putc(s, 8 * ETH_ALEN); /* Mac Addr Len - bits */
 		stream_put(s, evp->prefix.mac.octet, 6); /* Mac Addr */
 		stream_putc(s, 8 * ipa_len);		 /* IP address Length */
@@ -3566,7 +3579,7 @@ void bgp_evpn_encode_prefix(struct stream *s, struct prefix *p,
 	case BGP_EVPN_IMET_ROUTE:
 		stream_putc(s, 17); // TODO: length - assumes IPv4 address
 		stream_put(s, prd->val, 8);      /* RD */
-		stream_putl(s, 0);		 /* Ethernet Tag ID */
+		stream_putl(s, evp->prefix.eth_tag);		 /* Ethernet Tag ID */
 		stream_putc(s, IPV4_MAX_BITLEN); /* IP address Length - bits */
 		/* Originating Router's IP Addr */
 		stream_put_in_addr(s, &evp->prefix.ip.ipaddr_v4);
diff --git a/bgpd/bgp_evpn_vty.c b/bgpd/bgp_evpn_vty.c
index 2af151cc5550..0fbb5b11ec3c 100644
--- a/bgpd/bgp_evpn_vty.c
+++ b/bgpd/bgp_evpn_vty.c
@@ -330,7 +330,7 @@ static void bgp_evpn_show_route_header(struct vty *vty, struct bgp *bgp,
 		"* valid, > best, i - internal\n");
 	vty_out(vty, "Origin codes: i - IGP, e - EGP, ? - incomplete\n");
 	vty_out(vty,
-		"EVPN type-2 prefix: [2]:[ESI]:[EthTag]:[MAClen]:[MAC]:[IPlen]:[IP]\n");
+		"EVPN type-2 prefix: [2]:[EthTag]:[MAClen]:[MAC]:[IPlen]:[IP]\n");
 	vty_out(vty, "EVPN type-3 prefix: [3]:[EthTag]:[IPlen]:[OrigIP]\n\n");
 	vty_out(vty, "%s", ri_header);
 }
@@ -1978,7 +1978,7 @@ static void evpn_show_route_rd(struct vty *vty, struct bgp *bgp,
 			/* RD header and legend - once overall. */
 			if (rd_header && !json) {
 				vty_out(vty,
-					"EVPN type-2 prefix: [2]:[ESI]:[EthTag]:[MAClen]:[MAC]\n");
+					"EVPN type-2 prefix: [2]:[EthTag]:[MAClen]:[MAC]\n");
 				vty_out(vty,
 					"EVPN type-3 prefix: [3]:[EthTag]:[IPlen]:[OrigIP]\n");
 				vty_out(vty,
-- 
2.17.0

